### ISSUE TITLE
args doesn't always have len()

### ISSUE THREAD

**Issue by ahundt** [jrl-umi3218/Eigen3ToPython#20]
many functions check len(args) but `args` may not always support `len()`.

Therefore I suggest the first line of each such function do something like the following on the first line:
```python
if args is not None and hasattr(args, '__len__') and len(args):
    # current code here
else:
   raise TypeError('Unsupported <classname> constructor argument of type ' + str(type(args)))
```

Alternatively we could use **kwargs dictionary instead of args to avoid the problem entirely.

I think this safety is worth the overhead, and in the case it isn't the pure c class should most likely be used or a c++ version of your function should be implemented anyway. What do you think?

**Comment by haudren** [comment]
In which case does `args` have no `len` ? According to the Python (possibly incomplete) [docs](https://docs.python.org/3.5/reference/compound_stmts.html#function-definitions)
```
If the form “*identifier” is present, it is initialized to a tuple receiving any excess positional parameters, defaulting to the empty tuple. If the form “**identifier” is present, it is initialized to a new dictionary receiving any excess keyword arguments, defaulting to a new empty dictionary. Parameters after “*” or “*identifier” are keyword-only parameters and may only be passed used keyword arguments.
```

AFAIK, a tuple always has a len.

**Comment by ahundt** [comment]
> AFAIK, a tuple always has a len.

@haudren The user might not call the constructor with a tuple:

```python
$ import eigen as e
athundt@Andrews-2013-MacBook-Pro-3 ~/source/robonetracker/build vrepplug
$ e.Quaterniond(1)
xonsh: For full traceback set: $XONSH_SHOW_TRACEBACK = True
TypeError: <class 'int'>
```

**Comment by ahundt** [comment]
For some reason in my original post I decided to print instead of `raise TypeError(...)`, I've corrected that.

**Comment by gergondet** [comment]
@ahundt This has nothing to do with the fact that args is not a tuple (it is)

The culprit is actually
```python
"Unsupported argument types passed to Quaterniond ctor: ".join([str(type(x)) for x in args]
```
which I didn't catch in the review but should be:
```python
raise TypeError("Unsupported argument types passed to Quaterniond ctor: " + ", ".join([str(type(x)) for x in args] ))
```
which will give you a much better error message:
```python
TypeError: Unsupported argument types passed to Quaterniond ctor: <type 'int'>
```

**Comment by haudren** [comment]
Thanks Pierre, I couldn't see the bug :)

**Comment by haudren** [comment]
Fixed by c2a38ef

The output is now:
```
Traceback (most recent call last):
  File "test.py", line 3, in <module>
    e.Quaterniond(1)
  File "eigen/eigen.pyx", line 2593, in eigen.eigen.Quaterniond.__cinit__ (eigen/eigen.cpp:86944)
TypeError: Unsupported argument types passed to Quaterniond ctor: <type 'int'>

```

**Comment by ahundt** [comment]
Closed a bit too early! There are a lot more instances of similar strings that need to be fixed, all the Matrix classes, AngleAxis, etc.

**Comment by gergondet** [comment]
I forgot to post a message here but I fixed the other in d095fb1

**Comment by ahundt** [comment]
sweet, sorry for creating the bug in the first place. I think now that it works this error message is quite helpful, thanks for the fixes! :-)
